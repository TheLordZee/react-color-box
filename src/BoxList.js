import React, {useState} from "react"
import Box from "./Box"
import {v4 as uuid} from 'uuid';
import BoxForm from "./BoxForm";
import './BoxList.css'

const BoxList = () => {
    const INITIAL_STATE = {
        height: "",
        width: "",
        color: ""
    }
    const [list, setList] = useState([])
    const [formData, setFormData] = useState(INITIAL_STATE)

    const addBox = (boxData) => {
        const box = {...boxData, id: uuid()}
        setList(items => [...items, box])
    }

    const removeBox = (id) =>{
        let newList = [...list]
        for(let i = 0; i < newList.length; i++){
            if(newList[i].id === id){
                newList.splice(i,1)
            }
        }
        setList(newList)
    }

    const onChange = e => {
        let {name, value} = e.target;
        if(name === "height" || name === "width"){
            value = parseInt(value)
        }
        setFormData(fD => ({
            ...fD,
            [name]: value
        }))
    }

    const handleSubmit = e => {
        e.preventDefault();
        console.log(formData)
        addBox(formData);
        
        setFormData(INITIAL_STATE);
    }

    return(
        <div className="BoxList">
            <BoxForm handleSubmit={handleSubmit} onChange={onChange} formData={formData}/>
            <div className="BoxList-list">
                {list.map(
                    b => <Box 
                        key={b.id}
                        id={b.id}
                        width={b.width}
                        height={b.height} 
                        bgColor={b.color} 
                        removeBox={removeBox}
                    />
                )}
            </div>
        </div>
    )
}

export default BoxList;