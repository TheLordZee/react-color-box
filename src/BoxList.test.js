import React from "react";
import { fireEvent, render, screen } from '@testing-library/react';
import BoxList from "./BoxList"

it('renders', () => {
  render(<BoxList />);
});

it("matches snapshot", () => {
    const {asFragment} = render(<BoxList />);
    expect(asFragment()).toMatchSnapshot();
})
it('can add a new box', () => {
    const {getByLabelText, queryByText, container} = render(<BoxList />);

    const heightInput = getByLabelText("Height:")
    const widthInput = getByLabelText("Width:")
    const colorInput = getByLabelText("Color:")
    const submitBtn = queryByText("Add Box")

    fireEvent.change(heightInput, {target: {value: "20"}})
    fireEvent.change(widthInput, {target: {value: "20"}})
    fireEvent.change(colorInput, {target: {value: "blue"}})
    fireEvent.click(submitBtn)

    expect(container.getElementsByClassName('Box')[0]).toBeInTheDocument();
})

it('can delete a box', () => {
    const {debug, getByLabelText, queryByText, container} = render(<BoxList />);

    const heightInput = getByLabelText("Height:")
    const widthInput = getByLabelText("Width:")
    const colorInput = getByLabelText("Color:")
    const submitBtn = queryByText("Add Box")
    
    fireEvent.change(heightInput, {target: {value: "20"}})
    fireEvent.change(widthInput, {target: {value: "20"}})
    fireEvent.change(colorInput, {target: {value: "blue"}})
    fireEvent.click(submitBtn)
    let box = container.getElementsByClassName('Box')[0];
    
    expect(box).toBeInTheDocument();

    const deleteBtn = queryByText("X")
    fireEvent.click(deleteBtn);
    expect(box).not.toBeInTheDocument();
})