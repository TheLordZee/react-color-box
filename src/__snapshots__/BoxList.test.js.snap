// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="BoxList"
  >
    <form
      class="BoxForm"
    >
      <label
        for="height"
      >
        Height:
      </label>
      <input
        class="BoxForm-input"
        id="height"
        name="height"
        type="number"
        value=""
      />
      <label
        for="width"
      >
        Width:
      </label>
      <input
        class="BoxForm-input"
        id="width"
        name="width"
        type="number"
        value=""
      />
      <label
        for="color"
      >
        Color:
      </label>
      <input
        class="BoxForm-input"
        id="color"
        name="color"
        type="text"
        value=""
      />
      <br />
      <button
        class="BoxForm-btn"
      >
        Add Box
      </button>
    </form>
    <div
      class="BoxList-list"
    />
  </div>
</DocumentFragment>
`;
